rpc/blockchain.go

```diff
@@ -0,0 +1,664 @@
+	ctx context.Context, blockNumber BlockNumber,
+) ([]string, error) {
+	// Process arguments based on version
+	blockNum := blockNumber.EthBlockNumber()
+
+	// Ensure correct block
+	if blockNum.Int64() == 0 || blockNum.Int64() >= s.hmy.CurrentBlock().Number().Int64() {
+		return []string{}, nil
+	}
+	if isBlockGreaterThanLatest(s.hmy, blockNum) {
+		return nil, ErrRequestedBlockTooHigh
+	}
+
+	// Fetch signers
+	slots, mask, err := s.hmy.GetBlockSigners(ctx, blockNum)
+	if err != nil {
+		return nil, err
+	}
+
+	// Response output is the same for all versions
+	signers := []string{}
+	for _, validator := range slots {
+		oneAddress, err := internal_common.AddressToBech32(validator.EcdsaAddress)
+		if err != nil {
+			return nil, err
+		}
+		if ok, err := mask.KeyEnabled(validator.BLSPublicKey); err == nil && ok {
+			signers = append(signers, oneAddress)
+		}
+	}
+	return signers, nil
+}
+
+// GetBlockSignerKeys returns bls public keys that signed the block.
+func (s *PublicBlockchainService) GetBlockSignerKeys(
+	ctx context.Context, blockNumber BlockNumber,
+) ([]string, error) {
+	// Process arguments based on version
+	blockNum := blockNumber.EthBlockNumber()
+
+	// Ensure correct block
+	if blockNum.Int64() == 0 || blockNum.Int64() >= s.hmy.CurrentBlock().Number().Int64() {
+		return []string{}, nil
+	}
+	if isBlockGreaterThanLatest(s.hmy, blockNum) {
+		return nil, ErrRequestedBlockTooHigh
+	}
+
+	// Fetch signers
+	slots, mask, err := s.hmy.GetBlockSigners(ctx, blockNum)
+	if err != nil {
+		return nil, err
+	}
+
+	// Response output is the same for all versions
+	signers := []string{}
+	for _, validator := range slots {
+		if ok, err := mask.KeyEnabled(validator.BLSPublicKey); err == nil && ok {
+			signers = append(signers, validator.BLSPublicKey.Hex())
+		}
+	}
+	return signers, nil
+}
+
+// IsBlockSigner returns true if validator with address signed blockNum block.
+func (s *PublicBlockchainService) IsBlockSigner(
+	ctx context.Context, blockNumber BlockNumber, address string,
+) (bool, error) {
+	// Process arguments based on version
+	blockNum := blockNumber.EthBlockNumber()
+
+	// Ensure correct block
+	if blockNum.Int64() == 0 {
+		return false, nil
+	}
+	if isBlockGreaterThanLatest(s.hmy, blockNum) {
+		return false, ErrRequestedBlockTooHigh
+	}
+
+	// Fetch signers
+	slots, mask, err := s.hmy.GetBlockSigners(ctx, blockNum)
+	if err != nil {
+		return false, err
+	}
+
+	// Check if given address is in slots (response output is the same for all versions)
+	for _, validator := range slots {
+		oneAddress, err := internal_common.AddressToBech32(validator.EcdsaAddress)
+		if err != nil {
+			return false, err
+		}
+		if oneAddress != address {
+			continue
+		}
+		if ok, err := mask.KeyEnabled(validator.BLSPublicKey); err == nil && ok {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// GetSignedBlocks returns how many blocks a particular validator signed for last defaultBlocksPeriod (3 hours ~ 1500 blocks).
```

Comment:
why 1500 blocks as default?

Reply:
Think this is an arbitrary number that we have. The comment is actually wrong, its 15k blocks, which is around an epoch's worth of blocks. Therefore I'll make `defaultBlocksPeriod` the blocks per epoch from the config.